eth/executionclient/multi_client.go

```diff
@@ -0,0 +1,476 @@
+	currClientIdx   int
+	closed          chan struct{}
+}
+
+// NewMulti creates a new instance of MultiClient.
+func NewMulti(ctx context.Context, nodeAddrs []string, contractAddr ethcommon.Address, opts ...OptionMulti) (*MultiClient, error) {
+	if len(nodeAddrs) == 0 {
+		return nil, fmt.Errorf("no node address provided")
+	}
+
+	multiClient := &MultiClient{
+		nodeAddrs:                   nodeAddrs,
+		contractAddress:             contractAddr,
+		logger:                      zap.NewNop(),
+		followDistance:              DefaultFollowDistance,
+		connectionTimeout:           DefaultConnectionTimeout,
+		reconnectionInitialInterval: DefaultReconnectionInitialInterval,
+		reconnectionMaxInterval:     DefaultReconnectionMaxInterval,
+		logBatchSize:                DefaultHistoricalLogsBatchSize,
+	}
+
+	for _, opt := range opts {
+		opt(multiClient)
+	}
+
+	logger := multiClient.logger.WithOptions(zap.WithFatalHook(zapcore.WriteThenNoop))
+
+	for _, nodeAddr := range nodeAddrs {
+		singleClient, err := New(
+			ctx,
+			nodeAddr,
+			contractAddr,
+			WithLogger(logger),
+			WithFollowDistance(multiClient.followDistance),
+			WithConnectionTimeout(multiClient.connectionTimeout),
+			WithReconnectionInitialInterval(multiClient.reconnectionInitialInterval),
+			WithReconnectionMaxInterval(multiClient.reconnectionMaxInterval),
+			WithHealthInvalidationInterval(multiClient.healthInvalidationInterval),
+			WithSyncDistanceTolerance(multiClient.syncDistanceTolerance),
+		)
+		if err != nil {
+			return nil, fmt.Errorf("create single client: %w", err)
+		}
+
+		multiClient.clients = append(multiClient.clients, singleClient)
+	}
+
+	same, err := multiClient.assertSameChainIDs(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("assert same chain IDs: %w", err)
+	}
+	if !same {
+		return nil, fmt.Errorf("execution clients' chain IDs are not same")
+	}
+
+	return multiClient, nil
+}
+
+// assertSameChainIDs checks if all healthy clients have the same chain ID.
+// It sets firstChainID to the chain ID of the first healthy client encountered.
+func (ec *MultiClient) assertSameChainIDs(ctx context.Context) (bool, error) {
+	for i, client := range ec.clients {
+		addr := ec.nodeAddrs[i]
+
+		chainID, err := client.ChainID(ctx)
+		if err != nil {
+			ec.logger.Error("failed to get chain ID", zap.String("address", addr), zap.Error(err))
+			return false, fmt.Errorf("get chain ID: %w", err)
+		}
+		if ec.chainID == nil {
+			ec.chainID = chainID
+			continue
+		}
+		if ec.chainID.Cmp(chainID) != 0 {
+			ec.logger.Error("chain ID mismatch",
+				zap.String("observed_chain_id", ec.chainID.String()),
+				zap.String("checked_chain_id", chainID.String()),
+				zap.String("address", addr))
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+// FetchHistoricalLogs retrieves historical logs emitted by the contract starting from fromBlock.
+func (ec *MultiClient) FetchHistoricalLogs(ctx context.Context, fromBlock uint64) (<-chan BlockLogs, <-chan error, error) {
```

Comment:
Just fetching from a single client and then starting from scratch if is failed might be a smarter solution here. Since we sync history only on startup and its not a heavy long process, we should go with a simpler solution.